---
title: "ECHS1 KO RNA-seq"
author: "Mark Ziemann"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    fig_width: 7
    fig_height: 7
theme: cosmo
---

Source: https://github.com/markziemann/echs1_rnaseq

# Introduction

The samples are labelled as CON (control) and KO for knockout, ut are the untreated samples and 
dNs are the treated samples (mito biogenesis stimulator).

```{r,packages}

suppressPackageStartupMessages({
    library("zoo")
    library("tidyverse")
    library("reshape2")
    library("DESeq2")
    library("gplots")
    library("fgsea")
    library("MASS")
    library("mitch")
    library("eulerr")
    library("limma")
    library("topconfects")
    library("kableExtra")
    library("vioplot")
    library("beeswarm")
})

```

## Import read counts

Importing RNA-seq data

```{r,importdata}

tmp <- read.table("3col.tsv.gz",header=F)
x <- as.matrix(acast(tmp, V2~V1, value.var="V3", fun.aggregate = sum))
x <- as.data.frame(x)
accession <- sapply((strsplit(rownames(x),"\\|")),"[[",2)
symbol<-sapply((strsplit(rownames(x),"\\|")),"[[",6)
x$geneid <- paste(accession,symbol)
xx <- aggregate(. ~ geneid,x,sum)
rownames(xx) <- xx$geneid
xx$geneid = NULL
xx <- round(xx)
xx <- xx[,which(colnames(xx)!="test")]
xx[1:6,1:6]
dim(xx)

```

Fix the sample names.

They are duplicated for lane 1 and 2, which I will aggregate.

```{r,colnames}

colnames(xx) <- sapply(strsplit(colnames(xx),"_RNA_"),"[[",1)
colnames(xx) <- sapply(strsplit(sub("_","@",colnames(xx)),"@"),"[[",2)
labels <- unique(sapply(strsplit(colnames(xx),"\\."),"[[",1))
l <- lapply(labels,function(x) { rowSums(xx[,grep(x,colnames(xx))]) } )
ll <- do.call(cbind,l)
colnames(ll) <- labels
ll <- as.data.frame(ll[,order(colnames(ll))])
write.table(ll,file="counts.tsv",sep="\t",quote=FALSE)

rpm <- ll/colSums(ll) *1e6
write.table(rpm,file="rpm.tsv",sep="\t",quote=FALSE)

```

Make a sample sheet.
This will be easy because the groups are specified in the name.

```{r,samplesheet}

ss <- as.data.frame(colnames(ll))
ss$ko <- factor(grepl("KO", colnames(ll))) 
ss$trt <- factor(grepl("dNs",colnames(ll)))
rownames(ss) <- ss[,1]
ss[,1]=NULL
# can also set some colours for later
ss$cols <- c(rep("gray",4),rep("lightblue",4),rep("pink",4),
  rep("orange",4))

ss %>% kbl(caption = "Sample sheet") %>%
  kable_paper("hover", full_width = F)

```

## QC analysis

Here I'll look at a few different quality control measures.

```{r,qc1,fig.height=7,fig.width=7}

par(mar=c(5,8,3,1))
barplot(colSums(ll),horiz=TRUE,las=1,xlab="num reads",col=ss$cols)
sums <- colSums(ll)
sums <- sums[order(sums)]
barplot(sums,horiz=TRUE,las=1,xlab="num reads")
abline(v=15000000,col="red")

```

## MDS plot for all samples

Multidimensional scaling plot to show the variation between all samples, very similar to PCA.

Firstly with the data before aggregating technical replicates.

```{r,mds1,fig.height=7,fig.width=7}

mds <- cmdscale(dist(t(xx)))

plot(mds, xlab="Coordinate 1", ylab="Coordinate 2", 
  type = "p",bty="n",pch=19, cex=4 ,col="gray")
text(mds, labels=rownames(mds) )

```

Now for the data after aggregation.

It looks like the drug has a bigger effect than the gene KO.

```{r,mds2,fig.height=7,fig.width=7}

mds <- cmdscale(dist(t(ll)))

plot(mds, xlab="Coordinate 1", ylab="Coordinate 2", 
  type = "p",bty="n",pch=19, cex=4 ,col=ss$cols)
text(mds, labels=rownames(mds) )

pdf("mds1.pdf")

plot(mds, xlab="Coordinate 1", ylab="Coordinate 2",
  type = "p",bty="n",pch=19, cex=4 ,col=ss$cols)
text(mds, labels=rownames(mds) )

dev.off()

```

## Correlation heatmap

```{r,cor1,fig.height=7,fig.width=7}

heatmap.2(cor(ll),trace="n",main="Pearson correlation heatmap",margin=c(8,8))

pdf("cor1.pdf")
heatmap.2(cor(ll),trace="n",main="Pearson correlation heatmap",margin=c(8,8))
dev.off()

```

## MDS plot for UT samples

```{r,mds3,fig.height=7,fig.width=7}

ll2 <- ll[,grep("ut",colnames(ll))]

mds <- cmdscale(dist(t(ll2)))

plot(mds, xlab="Coordinate 1", ylab="Coordinate 2",
  type = "p",bty="n",pch=19, cex=4 ,col=ss$cols)
text(mds, labels=rownames(mds) )

pdf("mds2.pdf")
plot(mds, xlab="Coordinate 1", ylab="Coordinate 2",
  type = "p",bty="n",pch=19, cex=4 ,col=ss$cols)
text(mds, labels=rownames(mds) )
dev.off()

```

## MDS plot for dNs samples

```{r,mds4,fig.height=7,fig.width=7}

ll3 <- ll[,grep("dNs",colnames(ll))]

mds <- cmdscale(dist(t(ll3)))

plot(mds, xlab="Coordinate 1", ylab="Coordinate 2",
  type = "p",bty="n",pch=19, cex=4 ,col=ss$cols)
text(mds, labels=rownames(mds) )

pdf("mds3.pdf")
plot(mds, xlab="Coordinate 1", ylab="Coordinate 2",
  type = "p",bty="n",pch=19, cex=4 ,col=ss$cols)
text(mds, labels=rownames(mds) )
dev.off()

```


## Correlation heatmaps for UT and dNs separately

```{r,cor2,fig.height=7,fig.width=7}

heatmap.2(cor(ll2),trace="n",main="Pearson correlation heatmap",margin=c(8,8))

heatmap.2(cor(ll3),trace="n",main="Pearson correlation heatmap",margin=c(8,8))

pdf("cor2.pdf")
heatmap.2(cor(ll2),trace="n",main="Pearson correlation heatmap",margin=c(8,8))
dev.off()

pdf("cor3.pdf")
heatmap.2(cor(ll3),trace="n",main="Pearson correlation heatmap",margin=c(8,8))
dev.off()

```

## Set up the different datasets for differential expression analysis

Don't forget to remove poorly detected genes from the matrix with a threshold of 10 reads per sample on average.

There are 4 contrasts to set up.

1. Effect of the KO in untreated cells

2. Effect of the KO in treated cells

3. Effect of the drug in WT cells

4. Effect of the drug in KO cells

5. CON ut vs KO dNs. (what was altered in ECHS1 KO, that has been returned 
to normal levels (and no longer significantly different) to CON ut).
 
```{r,filter}

ss1 <- ss[which(ss$trt=="FALSE"),]
xx1 <- ll[which(colnames(ll) %in% rownames(ss1))]
xx1 <- xx1[which(rowSums(xx1)>=10),]
rpm1 <- xx1/colSums(xx1) *1e6

ss2 <- ss[which(ss$trt=="TRUE"),]
xx2 <- ll[which(colnames(ll) %in% rownames(ss2))]
xx2 <- xx2[which(rowSums(xx2)>=10),]
rpm2 <- xx2/colSums(xx2) *1e6

ss3 <- ss[which(ss$ko=="FALSE"),]
xx3 <- ll[which(colnames(ll) %in% rownames(ss3))]
xx3 <- xx3[which(rowSums(xx3)>=10),]
rpm3 <- xx3/colSums(xx3) *1e6

ss4 <- ss[which(ss$ko=="TRUE"),]
xx4 <- ll[which(colnames(ll) %in% rownames(ss4))]
xx4 <- xx4[which(rowSums(xx4)>=10),]
rpm4 <- xx4/colSums(xx4) *1e6

ss5 <- subset(ss,ko == trt)
xx5 <- ll[which(colnames(ll) %in% rownames(ss5))]
xx5 <- xx5[which(rowSums(xx5)>=10),]
rpm5 <- xx5/colSums(xx5) *1e6

```

## Differential expression with DESeq2

```{r,de01}

dds <- DESeqDataSetFromMatrix(countData = xx1 , colData = ss1, 
  design = ~ ko )
res <- DESeq(dds)
z<- results(res)
vsd <- vst(dds, blind=FALSE)
zz <- cbind(as.data.frame(z),assay(vsd))
dge <- as.data.frame(zz[order(zz$pvalue),])
dge[1:20,1:6] %>% 
  kbl(caption = "Top gene expression differences for contrast 1: Effect of the KO in untreated cells") %>% 
  kable_paper("hover", full_width = F)
dge1 <- dge
d1up <- rownames(subset(dge1,padj <= 0.05 & log2FoldChange > 0))
d1dn <- rownames(subset(dge1,padj <= 0.05 & log2FoldChange < 0))
write.table(dge1,file="dge1.tsv",quote=FALSE,sep="\t")

```

```{r,de02}

dds <- DESeqDataSetFromMatrix(countData = xx2 , colData = ss2, 
  design = ~ ko )
res <- DESeq(dds)
z<- results(res)
vsd <- vst(dds, blind=FALSE)
zz <- cbind(as.data.frame(z),assay(vsd))
dge <- as.data.frame(zz[order(zz$pvalue),])
dge[1:20,1:6] %>%
  kbl(caption = "Top gene expression differences for contrast 2: Effect of the KO in treated cells") %>%
  kable_paper("hover", full_width = F)
dge2 <- dge
d2up <- rownames(subset(dge2,padj <= 0.05 & log2FoldChange > 0))
d2dn <- rownames(subset(dge2,padj <= 0.05 & log2FoldChange < 0))
write.table(dge2,file="dge2.tsv",quote=FALSE,sep="\t")

```

```{r,de03}

dds <- DESeqDataSetFromMatrix(countData = xx3 , colData = ss3,
  design = ~ trt )
res <- DESeq(dds)
z<- results(res)
vsd <- vst(dds, blind=FALSE)
zz <- cbind(as.data.frame(z),assay(vsd))
dge <- as.data.frame(zz[order(zz$pvalue),])
dge[1:20,1:6] %>%
  kbl(caption = "Top gene expression differences for contrast 3: Effect of the drug in WT cells") %>%
  kable_paper("hover", full_width = F)
dge3 <- dge
d3up <- rownames(subset(dge3,padj <= 0.05 & log2FoldChange > 0))
d3dn <- rownames(subset(dge3,padj <= 0.05 & log2FoldChange < 0))
write.table(dge3,file="dge3.tsv",quote=FALSE,sep="\t")

```

```{r,de04}

dds <- DESeqDataSetFromMatrix(countData = xx4 , colData = ss4,
  design = ~ trt )
res <- DESeq(dds)
z<- results(res)
vsd <- vst(dds, blind=FALSE)
zz <- cbind(as.data.frame(z),assay(vsd))
dge <- as.data.frame(zz[order(zz$pvalue),])
dge[1:20,1:6] %>%
  kbl(caption = "Top gene expression differences for contrast 4: Effect of the drug in KO cells") %>%
  kable_paper("hover", full_width = F)
dge4 <- dge
d4up <- rownames(subset(dge4,padj <= 0.05 & log2FoldChange > 0))
d4dn <- rownames(subset(dge4,padj <= 0.05 & log2FoldChange < 0))
write.table(dge4,file="dge4.tsv",quote=FALSE,sep="\t")

```

```{r,de05}

dds <- DESeqDataSetFromMatrix(countData = xx5 , colData = ss5,
  design = ~ trt )
res <- DESeq(dds)
z<- results(res)
vsd <- vst(dds, blind=FALSE)
zz <- cbind(as.data.frame(z),assay(vsd))
dge <- as.data.frame(zz[order(zz$pvalue),])
dge[1:20,1:6] %>%
  kbl(caption = "Top gene expression differences for contrast 5: Does the drug restore changes caused by KO?") %>%
  kable_paper("hover", full_width = F)
dge5 <- dge
d5up <- rownames(subset(dge5,padj <= 0.05 & log2FoldChange > 0))
d5dn <- rownames(subset(dge5,padj <= 0.05 & log2FoldChange < 0))
write.table(dge5,file="dge5.tsv",quote=FALSE,sep="\t")

```

Here let's look at some plots.

MA plot shows the average level and fold change of all detected genes.
Volcano plot shows the fold change and the significance, as measured by -log(p-value).
Significant genes are shown as red points.

There are heatmaps of the top ranked genes by p-value.
Above the gene expression values there is a bar in yellow/orange colours. 
Yellow shows relatively low QuickDASH score and orange shows high score.

```{r,deplots1}

maplot <- function(de,contrast_name) {
  de <- de[which(!is.na(de$padj)),]
  sig <-subset(de, padj < 0.05 )
  up <-rownames(subset(de, padj < 0.05 & log2FoldChange > 0))
  dn <-rownames(subset(de, padj < 0.05 & log2FoldChange < 0))
  GENESUP <- length(up)
  GENESDN <- length(dn)
  DET=nrow(de)
  SUBHEADER = paste(GENESUP, "up, ", GENESDN, "down", DET, "detected")
  ns <-subset(de, padj > 0.05 )
  plot(log2(de$baseMean),de$log2FoldChange, 
       xlab="log2 basemean", ylab="log2 foldchange",
       pch=19, cex=0.5, col="dark gray",
       main=contrast_name, cex.main=1)
  points(log2(sig$baseMean),sig$log2FoldChange,
         pch=19, cex=0.5, col="red")
  mtext(SUBHEADER,cex = 1)
}

make_volcano <- function(de,name) {
    de <- de[which(!is.na(de$padj)),]
    de$pvalue[which(de$pvalue==0)] <- 1e-320
    sig <- subset(de,padj<0.05)
    N_SIG=nrow(sig)
    N_UP=nrow(subset(sig,log2FoldChange>0))
    N_DN=nrow(subset(sig,log2FoldChange<0))
    DET=nrow(de)
    HEADER=paste(N_SIG,"@5%FDR,", N_UP, "up", N_DN, "dn", DET, "detected")
    plot(de$log2FoldChange,-log10(de$pval),cex=0.5,pch=19,col="darkgray",
        main=name, xlab="log2 FC", ylab="-log10 pval")
    mtext(HEADER)
    grid()
    points(sig$log2FoldChange,-log10(sig$pval),cex=0.5,pch=19,col="red")
}

make_volcano2 <- function(de,name) {
    de <- de[which(!is.na(de$padj)),]
    de$pvalue[which(de$pvalue==0)] <- 1e-320
    sig <- subset(de,padj<0.05)
    N_SIG=nrow(sig)
    N_UP=nrow(subset(sig,log2FoldChange>0))
    N_DN=nrow(subset(sig,log2FoldChange<0))
    DET=nrow(de)
    HEADER=paste(N_SIG,"@5%FDR,", N_UP, "up", N_DN, "dn", DET, "detected")
    top <- head(sig,30)
    mylabels <- sapply(strsplit(rownames(top)," "),"[[",2)
    plot(de$log2FoldChange,-log10(de$pval),cex=0.5,pch=19,col="darkgray",
        main=name, xlab="log2 FC", ylab="-log10 pval")
    mtext(HEADER)
    grid()
    points(sig$log2FoldChange,-log10(sig$pval),cex=0.5,pch=19,col="red")
    text(top$log2FoldChange+0.2,-log10(top$pval),labels=mylabels, srt=35 ,cex=0.7)
}

make_heatmap <- function(de,name,myss,mx,n=30){
  colfunc <- colorRampPalette(c("blue", "white", "red"))
  mxn <- mx/rowSums(mx)*1000000
  x <- mxn[which(rownames(mxn) %in% rownames(head(de,n))),]
  heatmap.2(as.matrix(x),trace="none",col=colfunc(25),scale="row", margins = c(10,15), cexRow=0.7, 
    main=paste("Top ranked",n,"genes in",name) , ColSideColors = myss$cols  )
}

make_heatmap2 <- function(de,name,myss,mx,n=30){
  colfunc <- colorRampPalette(c("blue", "white", "red"))
  mxn <- mx/rowSums(mx)*1000000
  x <- mxn[which(rownames(mxn) %in% rownames(head(de,n))),]
  rownames(x) <- sapply(strsplit(rownames(x)," "),"[[",2)
  heatmap.2(as.matrix(x),trace="none",col=colfunc(25),scale="row", margins = c(10,15), cexRow=0.6,
    main=paste("Top ranked",n,"genes in",name) , ColSideColors = myss$cols  )
}

mitch_barplot <- function(res){
  sig <- head(subset(res$enrichment_result,p.adjustANOVA<0.05),30)
  sig <- sig[order(sig$s.dist),]
  par(mar=c(3,25,1,1)); barplot(sig$s.dist,horiz=TRUE,las=2,cex.names = 0.6,cex.axis = 0.6,
    names.arg=sig$set,main="Enrichment score") ;grid()
}


mitch_barplot2 <- function(res){
  sig_up <- head(subset(res$enrichment_result,p.adjustANOVA<0.05 & s.dist>0),15)
  sig_dn <- head(subset(res$enrichment_result,p.adjustANOVA<0.05 & s.dist<0),15)
  sig <- rbind(sig_up,sig_dn)
  sig <- sig[order(sig$s.dist),]
  par(mar=c(3,25,1,1)); barplot(sig$s.dist,horiz=TRUE,las=2,cex.names = 0.6,cex.axis = 0.6,
    names.arg=sig$set,main="Enrichment score") ;grid()
}

maplot(dge1,"Cont1: Effect of KO in untr cells")
make_volcano(dge1,"Cont1: Effect of KO in untr cells")
make_volcano2(dge1,"Cont1: Effect of KO in untr cells")
make_heatmap(dge1,"Cont1: Effect of KO in untr cells",ss1,xx1,n=50)
make_heatmap2(dge1,"Cont1: Effect of KO in untr cells",ss1,xx1,n=50)

maplot(dge2,"Cont2: Effect of KO in trt cells")
make_volcano(dge2,"Cont2: Effect of KO in trt cells")
make_volcano2(dge2,"Cont2: Effect of KO in trt cells")
make_heatmap(dge2,"Cont2: Effect of KO in trt cells",ss2,xx2,n=50)
make_heatmap2(dge2,"Cont2: Effect of KO in trt cells",ss2,xx2,n=50)

maplot(dge3,"Cont3: Effect of drug in WT cells")
make_volcano(dge3,"Cont3: Effect of drug in WT cells")
make_volcano2(dge3,"Cont3: Effect of drug in WT cells")
make_heatmap(dge3,"Cont3: Effect of drug in WT cells",ss3,xx3,n=50)
make_heatmap2(dge3,"Cont3: Effect of drug in WT cells",ss3,xx3,n=50)

maplot(dge4,"Cont4: Effect of drug in KO cells")
make_volcano(dge4,"Cont4: Effect of drug in KO cells")
make_volcano2(dge4,"Cont4: Effect of drug in KO cells")
make_heatmap(dge4,"Cont4: Effect of drug in KO cells",ss4,xx4,n=50)
make_heatmap2(dge4,"Cont4: Effect of drug in KO cells",ss4,xx4,n=50)

maplot(dge5,"CON ut vs KO dNs")
make_volcano(dge5,"CON ut vs KO dNs")
make_volcano2(dge5,"CON ut vs KO dNs")
make_heatmap(dge5,"CON ut vs KO dNs",ss5,xx5,n=50)
make_heatmap2(dge5,"CON ut vs KO dNs",ss5,xx5,n=50)

pdf("dge1.pdf")
maplot(dge1,"Cont1: Effect of KO in untr cells")
make_volcano(dge1,"Cont1: Effect of KO in untr cells")
make_volcano2(dge1,"Cont1: Effect of KO in untr cells")
make_heatmap(dge1,"Cont1: Effect of KO in untr cells",ss1,xx1,n=50)
make_heatmap2(dge1,"Cont1: Effect of KO in untr cells",ss1,xx1,n=50)
dev.off()

pdf("dge2.pdf")
maplot(dge2,"Cont2: Effect of KO in trt cells")
make_volcano(dge2,"Cont2: Effect of KO in trt cells")
make_volcano2(dge2,"Cont2: Effect of KO in trt cells")
make_heatmap(dge2,"Cont2: Effect of KO in trt cells",ss2,xx2,n=50)
make_heatmap2(dge2,"Cont2: Effect of KO in trt cells",ss2,xx2,n=50)
dev.off()

pdf("dge3.pdf")
maplot(dge3,"Cont3: Effect of drug in WT cells")
make_volcano(dge3,"Cont3: Effect of drug in WT cells")
make_volcano2(dge3,"Cont3: Effect of drug in WT cells")
make_heatmap(dge3,"Cont3: Effect of drug in WT cells",ss3,xx3,n=50)
make_heatmap2(dge3,"Cont3: Effect of drug in WT cells",ss3,xx3,n=50)
dev.off()

pdf("dge4.pdf")
maplot(dge4,"Cont4: Effect of drug in KO cells")
make_volcano(dge4,"Cont4: Effect of drug in KO cells")
make_volcano2(dge4,"Cont4: Effect of drug in KO cells")
make_heatmap(dge4,"Cont4: Effect of drug in KO cells",ss4,xx4,n=50)
make_heatmap2(dge4,"Cont4: Effect of drug in KO cells",ss4,xx4,n=50)
dev.off()

pdf("dge5.pdf")
maplot(dge5,"CON ut vs KO dNs")
make_volcano(dge5,"CON ut vs KO dNs")
make_volcano2(dge5,"CON ut vs KO dNs")
make_heatmap(dge5,"CON ut vs KO dNs",ss5,xx5,n=50)
make_heatmap2(dge5,"CON ut vs KO dNs",ss5,xx5,n=50)
dev.off()

```

## Heatmap of DEGs

Contrast 1 and 4 - get top 25 genes and make heatmap of all sample groups.

```{r,heat1}

g1 <- head(rownames(dge1),25)
g4 <- head(rownames(dge4),25)
g14 <- union(g1,g4)

fc1 <- dge1[which(rownames(dge1) %in% g14),"log2FoldChange",drop=F]

rpm14 <- rpm[which(rownames(rpm) %in% g14),]
rpm14 <- merge(rpm14,fc1,by=0)
rpm14 <- rpm14[order(rpm14$log2FoldChange),]
rpm14$log2FoldChange=NULL
rownames(rpm14) <- rpm14$Row.names
rpm14$Row.names=NULL
rownames(rpm14) <- sapply(strsplit(rownames(rpm14)," "),"[[",2)
rpm14 <- rpm14[,rev(order(colnames(rpm14)))]

# reorder columns
rpm14 <- rpm14[,c( grep("CON_ut",colnames(rpm14)) , grep("KO_ut",colnames(rpm14)) , grep("KO_dN",colnames(rpm14)) , grep("CON_dN",colnames(rpm14)) )]

colfunc <- colorRampPalette(c("blue", "white", "red"))

heatmap.2(as.matrix(rpm14),trace="none",col=colfunc(25),scale="row", 
  margins = c(8,15), cexRow=0.5 , cexCol=0.8)

heatmap.2(as.matrix(rpm14),trace="none",col=colfunc(25),scale="row", 
  Colv="none" , dendrogram="none",
  margins = c(8,15), cexRow=0.5 , cexCol=0.8)

pdf("heat1.pdf")

heatmap.2(as.matrix(rpm14),trace="none",col=colfunc(25),scale="row",
  margins = c(8,15), cexRow=0.5 , cexCol=0.8)

heatmap.2(as.matrix(rpm14),trace="none",col=colfunc(25),scale="row", 
  Colv="none" , dendrogram="none",
  margins = c(8,15), cexRow=0.5 , cexCol=0.8)

dev.off()

rpm14 <- rpm[which(rownames(rpm) %in% g14),]

rownames(rpm14) <- sapply(strsplit(rownames(rpm14)," "),"[[",2)

# reorder columns
rpm14 <- rpm14[,c( grep("CON_ut",colnames(rpm14)) , grep("KO_ut",colnames(rpm14)) , grep("KO_dN",colnames(rpm14)) , grep("CON_dN",colnames(rpm14)) )]

colfunc <- colorRampPalette(c("blue", "white", "red"))

heatmap.2(as.matrix(rpm14),trace="none",col=colfunc(25),scale="row", 
  margins = c(8,15), cexRow=0.5 , cexCol=0.8)

heatmap.2(as.matrix(rpm14),trace="none",col=colfunc(25),scale="row",
  Colv="none" , dendrogram="none",
  margins = c(8,15), cexRow=0.5 , cexCol=0.8)

pdf("heat2.pdf")

heatmap.2(as.matrix(rpm14),trace="none",col=colfunc(25),scale="row",
  margins = c(8,15), cexRow=0.5 , cexCol=0.8)

heatmap.2(as.matrix(rpm14),trace="none",col=colfunc(25),scale="row", 
  Colv="none" , dendrogram="none",
  margins = c(8,15), cexRow=0.5 , cexCol=0.8)

dev.off()

rpm1x <- rpm1[which(rownames(rpm1) %in% g14),]
fc1 <- dge1[which(rownames(dge1) %in% g14),"log2FoldChange",drop=F]
rpm1x <- merge(rpm1x,fc1,by=0)
rpm1x <- rpm1x[order(rpm1x$log2FoldChange),]
rpm1x$log2FoldChange=NULL
rownames(rpm1x) <- rpm1x$Row.names
rpm1x$Row.names=NULL
rownames(rpm1x) <- sapply(strsplit(rownames(rpm1x)," "),"[[",2)
rpm1x <- rpm1x[,rev(order(colnames(rpm1x)))]

rpm4x <- rpm4[which(rownames(rpm4) %in% g14),]
rpm4x <- merge(rpm4x,fc1,by=0)
rpm4x <- rpm4x[order(rpm4x$log2FoldChange),]
rpm4x$log2FoldChange=NULL
rownames(rpm4x) <- rpm4x$Row.names
rpm4x$Row.names=NULL
rownames(rpm4x) <- sapply(strsplit(rownames(rpm4x)," "),"[[",2)
rpm4x <- rpm4x[,rev(order(colnames(rpm4x)))]

heatmap.2(as.matrix(rpm1x),trace="none",col=colfunc(25),scale="row", margins = c(8,15), 
  cexRow=0.5 , cexCol=0.8, Rowv=FALSE, dendrogram='none',Colv=FALSE)

heatmap.2(as.matrix(rpm4x),trace="none",col=colfunc(25),scale="row", margins = c(8,15), 
  cexRow=0.5 , cexCol=0.8, Rowv=FALSE, dendrogram='none',Colv=FALSE)

pdf("heat3.pdf")

heatmap.2(as.matrix(rpm1x),trace="none",col=colfunc(25),scale="row", margins = c(8,15), 
  cexRow=0.5 , cexCol=0.8, Rowv=FALSE, dendrogram='none',Colv=FALSE)

heatmap.2(as.matrix(rpm4x),trace="none",col=colfunc(25),scale="row", margins = c(8,15), 
  cexRow=0.5 , cexCol=0.8, Rowv=FALSE, dendrogram='none',Colv=FALSE)
dev.off()

```

## Euler diagrams of DEGs

All DEGs

```{r,euler1}

par(cex.main=0.5)

par(mar=c(2,2,2,2))

v0 <- list("C1 up"=d1up,"C1 dn"=d1dn, "C2 up" = d2up, "C2 dn"=d2dn, "C3 up"=d3up, "C3 dn"=d3dn, "C4 up"=d4up, "C4 dn"=d4dn)

plot(euler(v0),quantities = TRUE, edges = "gray", main="DEGs")

pdf("euler1.pdf")
plot(euler(v0),quantities = TRUE, edges = "gray", main="DEGs")
dev.off()

```

Now with just the DEGs from contrasts 1 and 4.

```{r,euler2}

par(cex.main=0.5)

par(mar=c(2,2,2,2))

v0 <- list("KO up"=d1up,"KO dn"=d1dn, "drug up"=d4up, "drug dn"=d4dn)

plot(euler(v0),quantities = TRUE, edges = "gray", main="DEGs")

pdf("euler2.pdf")
plot(euler(v0),quantities = TRUE, edges = "gray", main="DEGs")
dev.off()

```

Now with just the DEGs from contrasts 1 and 5.

```{r,euler3}

par(cex.main=0.5)

par(mar=c(2,2,2,2))

v0 <- list("KO up"=d1up,"KO dn"=d1dn, "drugtrt up"=d5up, "drugtrt dn"=d5dn)

plot(euler(v0),quantities = TRUE, edges = "gray", main="DEGs")

pdf("euler3.pdf")
plot(euler(v0),quantities = TRUE, edges = "gray", main="DEGs")
dev.off()

```

## Single contrast pathway analysis with mitch

Firstly need to conduct mitch enrichment analysis for each contrast separately.

```{r,mitch1}

#download.file("https://reactome.org/download/current/ReactomePathways.gmt.zip", destfile="ReactomePathways.gmt.zip")
#unzip("ReactomePathways.gmt.zip")
genesets <- gmt_import("ReactomePathways.gmt")

# gene table
gt <- as.data.frame(rownames(xx))
gt$gene <- sapply(strsplit(gt[,1]," "),"[[",2)

m1 <- mitch_import(dge1, DEtype="deseq2",geneTable=gt)
mres1 <- mitch_calc(m1, genesets, priority="effect")
head(mres1$enrichment_result,20) %>%
  kbl(caption = "Top gene pathway differences in contrast 1") %>%
  kable_paper("hover", full_width = F)
mitch_barplot(mres1)
mitch_barplot2(mres1)
mitch_plots(mres1,outfile="mitch1.pdf")
m1top <- subset(mres1$enrichment_result,p.adjustANOVA<0.05)
m1up <- subset(m1top,s.dist>0)$set
m1dn <- subset(m1top,s.dist<0)$set
#mitch_report(mres1,outfile="mitch1.html",overwrite=TRUE)
write.table(mres1$enrichment_result,file="mitch1.tsv",quote=FALSE,sep="\t",row.names=FALSE)

m1top_up <- head(subset(m1top,s.dist>0),10)[,"s.dist"]
names(m1top_up) <- head(subset(m1top,s.dist>0),10)[,"set"]
m1top_dn <- head(subset(m1top,s.dist<0),10)[,"s.dist"]
names(m1top_dn) <- head(subset(m1top,s.dist<0),10)[,"set"]
m1top_updn <- c(m1top_up,m1top_dn)
m1top_updn <- m1top_updn[order(m1top_updn)]

pdf("f2b.pdf")
par(mar=c(5,25,3,1))
barplot(m1top_updn,horiz=TRUE,las=1,col="darkgray",xlab="Enrichment score",cex.names=0.8,xlim=c(-1,1))
grid()
dev.off()

m2 <- mitch_import(dge2, DEtype="deseq2",geneTable=gt)
mres2 <- mitch_calc(m2, genesets, priority="effect")
head(mres2$enrichment_result,20) %>% 
  kbl(caption = "Top gene pathway differences in contrast 2") %>%
  kable_paper("hover", full_width = F)
mitch_barplot(mres2)
mitch_barplot2(mres2)
mitch_plots(mres2,outfile="mitch2.pdf")
m2top <- subset(mres2$enrichment_result,p.adjustANOVA<0.05)
m2up <- subset(m2top,s.dist>0)$set
m2dn <- subset(m2top,s.dist<0)$set
#mitch_report(mres2,outfile="mitch2.html",overwrite=TRUE)
write.table(mres2$enrichment_result,file="mitch2.tsv",quote=FALSE,sep="\t",row.names=FALSE)

m3 <- mitch_import(dge3, DEtype="deseq2",geneTable=gt)
mres3 <- mitch_calc(m3, genesets, priority="effect")
head(mres3$enrichment_result,20) %>%
  kbl(caption = "Top gene pathway differences in contrast 3") %>%
  kable_paper("hover", full_width = F)
mitch_barplot(mres3)
mitch_barplot2(mres3)
mitch_plots(mres3,outfile="mitch3.pdf")
m3top <- subset(mres3$enrichment_result,p.adjustANOVA<0.05)
m3up <- subset(m3top,s.dist>0)$set
m3dn <- subset(m3top,s.dist<0)$set
#mitch_report(mres3,outfile="mitch3.html",overwrite=TRUE)
write.table(mres3$enrichment_result,file="mitch3.tsv",quote=FALSE,sep="\t",row.names=FALSE)

m4 <- mitch_import(dge4, DEtype="deseq2",geneTable=gt)
mres4 <- mitch_calc(m4, genesets, priority="effect")
head(mres4$enrichment_result,20) %>%
  kbl(caption = "Top gene pathway differences in contrast 4") %>%
  kable_paper("hover", full_width = F)
mitch_barplot(mres4)
mitch_barplot2(mres4)
mitch_plots(mres4,outfile="mitch4.pdf")
m4top <- subset(mres4$enrichment_result,p.adjustANOVA<0.05)
m4up <- subset(m4top,s.dist>0)$set
m4dn <- subset(m4top,s.dist<0)$set
#mitch_report(mres4,outfile="mitch4.html",overwrite=TRUE)
write.table(mres4$enrichment_result,file="mitch4.tsv",quote=FALSE,sep="\t",row.names=FALSE)

m5 <- mitch_import(dge5, DEtype="deseq2",geneTable=gt)
mres5 <- mitch_calc(m5, genesets, priority="effect")
head(mres5$enrichment_result,20) %>%
  kbl(caption = "Top gene pathway differences in contrast 5") %>%
  kable_paper("hover", full_width = F)
mitch_barplot(mres5)
mitch_barplot2(mres5)
mitch_plots(mres5,outfile="mitch5.pdf")
m5top <- subset(mres5$enrichment_result,p.adjustANOVA<0.05)
m5up <- subset(m5top,s.dist>0)$set
m5dn <- subset(m5top,s.dist<0)$set
#mitch_report(mres5,outfile="mitch5.html",overwrite=TRUE)
write.table(mres5$enrichment_result,file="mitch5.tsv",quote=FALSE,sep="\t",row.names=FALSE)

pdf("mitch_barplots.pdf")
mitch_barplot(mres1)
mitch_barplot2(mres1)
mitch_barplot(mres2)
mitch_barplot2(mres2)
mitch_barplot(mres3)
mitch_barplot2(mres3)
mitch_barplot(mres4)
mitch_barplot2(mres4)
mitch_barplot(mres5)
mitch_barplot2(mres5)
dev.off()

```

Make some ridge plots.

```{r,ridgeplots0}

myres <- list(mres1,mres2,mres3,mres4) 
z <- lapply(myres, function(mres) {
  sets <- head(mres$enrichment_result,10)
  sets <- sets[order(sets$s.dist),"set"]
  ridgedata <- lapply( sets , function(myset) {
    genes <- genesets[[which(names(genesets) == myset)]]
    generanks <- mres$ranked_profile[which(rownames(mres$ranked_profile) %in% genes),]
    return(generanks)
  })
  names(ridgedata) <- sets
  myrange <- range(mres$ranked_profile[,1])
  par(mar=c(5,20,3,2))
 
 vioplot(ridgedata,horizontal=TRUE,las=1,
    ylim=myrange,main="Top 10 genesets",
    xlab="differential gene rank")

 vioplot(ridgedata,horizontal=TRUE,las=1,
    ylim=myrange,col="white",main="Top 10 genesets",
    xlab="differential gene rank",
    rectCol="gray")
  beeswarm(ridgedata,add=TRUE,horizontal = TRUE, pch=19,cex=0.6,corral = "gutter")
})

pdf("ridge1.pdf")
z <- lapply(myres, function(mres) {
  sets <- head(mres$enrichment_result,10)
  sets <- sets[order(sets$s.dist),"set"]
  ridgedata <- lapply( sets , function(myset) {
    genes <- genesets[[which(names(genesets) == myset)]]
    generanks <- mres$ranked_profile[which(rownames(mres$ranked_profile) %in% genes),]
    return(generanks)
  })
  names(ridgedata) <- sets
  myrange <- range(mres$ranked_profile[,1])
  par(mar=c(5,20,3,2))

 vioplot(ridgedata,horizontal=TRUE,las=1,
    ylim=myrange,main="Top 10 genesets",
    xlab="differential gene rank")

 vioplot(ridgedata,horizontal=TRUE,las=1,
    ylim=myrange,col="white",main="Top 10 genesets",
    xlab="differential gene rank",
    rectCol="gray")
  beeswarm(ridgedata,add=TRUE,horizontal = TRUE, pch=19,cex=0.6,corral = "gutter")
})
dev.off()

```

Make some heatmaps

```{r,heatmaps}

sets <- head(mres1$enrichment_result$set,10)

z <- lapply(sets, function(set){
  genes <- genesets[[which(names(genesets) == set)]]
  refgenes <- sapply(strsplit(rownames(rpm1)," "),"[[",2)
  hm <- as.matrix(rpm1[which(refgenes %in% genes),])
  hm <- hm/rowMeans(hm)
  colfunc <- colorRampPalette(c("blue", "white", "red"))
  heatmap.2(hm,trace="none",col=colfunc(25),scale="row", margins = c(6,15), cexRow=0.9,
    cexCol=0.9, main=paste(set) )
})

z <- lapply(sets, function(set){
  genes <- genesets[[which(names(genesets) == set)]]
  refgenes <- sapply(strsplit(rownames(rpm)," "),"[[",2)
  hm <- as.matrix(rpm[which(refgenes %in% genes),])
  hm <- hm/rowMeans(hm)
  colfunc <- colorRampPalette(c("blue", "white", "red"))
  heatmap.2(hm,trace="none",col=colfunc(25),scale="row", margins = c(6,15), cexRow=0.7,
    cexCol=0.9, main=paste(set) )
})

pdf("peheat1.pdf")
z <- lapply(sets, function(set){
  genes <- genesets[[which(names(genesets) == set)]]
  refgenes <- sapply(strsplit(rownames(rpm1)," "),"[[",2)
  hm <- as.matrix(rpm1[which(refgenes %in% genes),])
  hm <- hm/rowMeans(hm)
  colfunc <- colorRampPalette(c("blue", "white", "red"))
  heatmap.2(hm,trace="none",col=colfunc(25),scale="row", margins = c(6,15), cexRow=0.9,
    cexCol=0.9, main=paste(set) )
})

z <- lapply(sets, function(set){
  genes <- genesets[[which(names(genesets) == set)]]
  refgenes <- sapply(strsplit(rownames(rpm)," "),"[[",2)
  hm <- as.matrix(rpm[which(refgenes %in% genes),])
  hm <- hm/rowMeans(hm)
  colfunc <- colorRampPalette(c("blue", "white", "red"))
  heatmap.2(hm,trace="none",col=colfunc(25),scale="row", margins = c(6,15), cexRow=0.7,
    cexCol=0.9, main=paste(set) )
})
dev.off()

```

Now an Euler diagram of differentially regulated gene sets from contrasts 1 and 4.

```{r,euler4}

par(cex.main=0.5)

par(mar=c(2,2,2,2))

v0 <- list("KO up"=m1up,"KO dn"=m1dn, "drug up"=m4up, "drug dn"=m4dn)

plot(euler(v0),quantities = TRUE, edges = "gray", main="Reactome pathways")

pdf("gseuler1.pdf")
plot(euler(v0),quantities = TRUE, edges = "gray", main="Reactome pathways")
dev.off()

```

## Multi contrast pathway analysis with mitch

I'm going to focus on the two contrasts which I think are the most important WRT mito disease; #1 and #4.
The effect of the KO, and the effect of the drug when the gene is knocked out.

```{r,mitch2}

mm <- list("KO"=dge1,"drug"=dge4)

m <- mitch_import(mm, DEtype="deseq2",geneTable=gt)
mres <- mitch_calc(m, genesets, priority="effect")
head(mres$enrichment_result,20) %>% 
kbl(caption = "Top gene pathway differences in contrast 1 and 4") %>% kable_paper("hover", full_width = F)
#mitch_report(mres,outfile="mitchmulti_effect.html",overwrite=TRUE)
mitch_plots(mres, outfile="mitchmulti_effect.pdf")
write.table(mres$enrichment_result,file="mitch_multi_effect.tsv",quote=FALSE,sep="\t",row.names=FALSE)
n=40
mx <- as.matrix(mres$enrichment_result[1:n,4:5])
rownames(mx) <- mres$set
rownames(mx) <- mres$enrichment_result$set[1:n]

colfunc <- colorRampPalette(c("blue", "white", "red"))

heatmap.2(mx,trace="none",col=colfunc(25),scale="none", margins = c(5,15), cexRow=0.7, 
    cexCol=0.9, main=paste("Top ranked Reactomes") )

```

Mitch can also be used to search for discordant pathways.
Gene sets are prioritised by standard deviation.

```{r,mitch3}

mm <- list("KO"=dge1,"drug"=dge4)

m <- mitch_import(mm, DEtype="deseq2",geneTable=gt)
mres <- mitch_calc(m, genesets, priority="SD")
head(mres$enrichment_result,20) %>% kbl(caption = "Top gene pathway differences in contrast 1 and 4") %>% kable_paper("hover", full_width = F)
#mitch_report(mres,outfile="mitchmulti_discord.html",overwrite=TRUE)
mitch_plots(mres, outfile="mitchmulti_discord.pdf")

write.table(mres$enrichment_result,file="mitch_multi_discord.tsv",quote=FALSE,sep="\t",row.names=FALSE)

n=50

mx <- as.matrix(mres$enrichment_result[1:n,4:5])
rownames(mx) <- mres$set
rownames(mx) <- mres$enrichment_result$set[1:n]

colfunc <- colorRampPalette(c("blue", "white", "red"))

heatmap.2(mx,trace="none",col=colfunc(25),scale="none", margins = c(5,25), cexRow=0.6,
    cexCol=0.9, main=paste("Top ranked Reactomes") )

pdf("mitchmulti_discord_heat.pdf")
heatmap.2(mx,trace="none",col=colfunc(25),scale="none", margins = c(5,25), cexRow=0.6,
    cexCol=0.9, main=paste("Top ranked Reactomes") )
dev.off()


```

## Custom gene sets

```{r,mitch_custom1}

custom_set_names <- readLines("custom_pathway_list.txt")
custom_sets <- genesets[which(names(genesets) %in% custom_set_names)]

m1 <- mitch_import(dge1, DEtype="deseq2",geneTable=gt)
mres1 <- mitch_calc(m1, custom_sets, priority="effect")
head(mres1$enrichment_result,20) %>% kbl(caption = "Top mito pathway differences in contrast 1") %>% kable_paper("hover", full_width = F)
m1top <- subset(mres1$enrichment_result,p.adjustANOVA<0.05)
m1up <- subset(m1top,s.dist>0)$set
m1dn <- subset(m1top,s.dist<0)$set
#mitch_report(mres1,outfile="mitch_mito1.html",overwrite=TRUE)
mitch_plots(mres1, outfile="mitch_mito1.pdf")
write.table(mres1$enrichment_result,file="mitch_mito1.tsv",quote=FALSE,sep="\t",row.names=FALSE)

m2 <- mitch_import(dge2, DEtype="deseq2",geneTable=gt)
mres2 <- mitch_calc(m2, custom_sets, priority="effect")
head(mres2$enrichment_result,20) %>% kbl(caption = "Top mito pathway differences in contrast 2") %>% kable_paper("hover", full_width = F)
m2top <- subset(mres2$enrichment_result,p.adjustANOVA<0.05)
m2up <- subset(m2top,s.dist>0)$set
m2dn <- subset(m2top,s.dist<0)$set
#mitch_report(mres2,outfile="mitch_mito2.html",overwrite=TRUE)
mitch_plots(mres2, outfile="mitch_mito2.pdf")
write.table(mres2$enrichment_result,file="mitch_mito2.tsv",quote=FALSE,sep="\t",row.names=FALSE)

m3 <- mitch_import(dge3, DEtype="deseq2",geneTable=gt)
mres3 <- mitch_calc(m3, custom_sets, priority="effect")
head(mres3$enrichment_result,20) %>% kbl(caption = "Top mito pathway differences in contrast 3") %>% kable_paper("hover", full_width = F)
m3top <- subset(mres3$enrichment_result,p.adjustANOVA<0.05)
m3up <- subset(m3top,s.dist>0)$set
m3dn <- subset(m3top,s.dist<0)$set
#mitch_report(mres3,outfile="mitch_mito3.html",overwrite=TRUE)
mitch_plots(mres3, outfile="mitch_mito3.pdf")
write.table(mres3$enrichment_result,file="mitch_mito3.tsv",quote=FALSE,sep="\t",row.names=FALSE)

m4 <- mitch_import(dge4, DEtype="deseq2",geneTable=gt)
mres4 <- mitch_calc(m4, custom_sets, priority="effect")
head(mres4$enrichment_result,20) %>% kbl(caption = "Top mito pathway differences in contrast 4") %>% kable_paper("hover", full_width = F)
m4top <- subset(mres4$enrichment_result,p.adjustANOVA<0.05)
m4up <- subset(m4top,s.dist>0)$set
m4dn <- subset(m4top,s.dist<0)$set
#mitch_report(mres4,outfile="mitch_mito4.html",overwrite=TRUE)
mitch_plots(mres4, outfile="mitch_mito4.pdf")
write.table(mres4$enrichment_result,file="mitch_mito4.tsv",quote=FALSE,sep="\t",row.names=FALSE)

m5 <- mitch_import(dge5, DEtype="deseq2",geneTable=gt)
mres5 <- mitch_calc(m5, custom_sets, priority="effect")
head(mres5$enrichment_result,20) %>% kbl(caption = "Top mito pathway differences in contrast 5") %>% kable_paper("hover", full_width = F)
m5top <- subset(mres5$enrichment_result,p.adjustANOVA<0.05)
m5up <- subset(m5top,s.dist>0)$set
m5dn <- subset(m5top,s.dist<0)$set
#mitch_report(mres5,outfile="mitch_mito5.html",overwrite=TRUE)
mitch_plots(mres5, outfile="mitch_mito5.pdf")
write.table(mres5$enrichment_result,file="mitch_mito5.tsv",quote=FALSE,sep="\t",row.names=FALSE)

```

Make some ridge plots.  

```{r,ridgeplots1}

myres <- list(mres1,mres2,mres3,mres4)
z <- lapply(myres, function(mres) {
  sets <- head(mres$enrichment_result,10)
  sets <- sets[order(sets$s.dist),"set"]
  ridgedata <- lapply( sets , function(myset) {
    genes <- genesets[[which(names(genesets) == myset)]]
    generanks <- mres$ranked_profile[which(rownames(mres$ranked_profile) %in% genes),]
    return(generanks)
  })
  names(ridgedata) <- sets
  myrange <- range(mres$ranked_profile[,1])
  par(mar=c(5,20,3,2))
  
  vioplot(ridgedata,horizontal=TRUE,las=1,
    ylim=myrange,main="Top 10 genesets",
    xlab="differential gene rank")

  vioplot(ridgedata,horizontal=TRUE,las=1,
    ylim=myrange,col="white",main="Top 10 genesets",
    xlab="differential gene rank",
    rectCol="gray")
  beeswarm(ridgedata,add=TRUE,horizontal = TRUE, pch=19,cex=0.6,corral = "gutter")
})


#Fig2c
myres <- mres1
  sets <- head(myres$enrichment_result,10)
  sets <- sets[order(sets$s.dist),"set"]
  ridgedata <- lapply( sets , function(myset) {
    genes <- genesets[[which(names(genesets) == myset)]]
    generanks <- myres$ranked_profile[which(rownames(myres$ranked_profile) %in% genes),]
    return(generanks)
  })
  names(ridgedata) <- sets
  myrange <- range(myres$ranked_profile[,1])

pdf("fig2C.pdf")
  par(mar=c(5,20,3,2))
  vioplot(ridgedata,horizontal=TRUE,las=1,
    ylim=myrange,main="Top 10 genesets",
    xlab="differential gene rank")
dev.off()

# now custom gene sets part 1
custom_set_names <- readLines("custom_pathway_list.txt")
myres <- list(mres1,mres2,mres3,mres4)
z <- lapply(myres, function(mres) {
  sets <- custom_set_names
  mres$enrichment_result <- mres$enrichment_result[which(mres$enrichment_result$set %in% sets),]
  mres$enrichment_result <- mres$enrichment_result[order(mres$enrichment_result$s.dist),]
  sets <- mres$enrichment_result$set
  ridgedata <- lapply( sets , function(myset) {
    genes <- genesets[[which(names(genesets) == myset)]]
    generanks <- mres$ranked_profile[which(rownames(mres$ranked_profile) %in% genes),]
    return(generanks)
  })
  names(ridgedata) <- sets
  myrange <- range(mres$ranked_profile[,1])
  par(mar=c(5,20,3,2))

  vioplot(ridgedata,horizontal=TRUE,las=1,
    ylim=myrange,main="Gene sets of interest",
    xlab="differential gene rank")


  vioplot(ridgedata,horizontal=TRUE,las=1,
    ylim=myrange,col="white",main="Gene sets of interest",
    rectCol="gray", xlab="differential gene rank")
  beeswarm(ridgedata,add=TRUE,horizontal = TRUE, pch=19,cex=0.6,corral = "gutter")
})

# now custom gene sets part 2
custom_set_names2 <- readLines("custom_pathway_list2.txt")
myres <- list(mres1,mres2,mres3,mres4)
z <- lapply(myres, function(mres) {
  sets <- custom_set_names2[which( custom_set_names2 %in% names(custom_sets))]
  mres$enrichment_result <- mres$enrichment_result[which(mres$enrichment_result$set %in% sets),]
  mres$enrichment_result <- mres$enrichment_result[order(mres$enrichment_result$s.dist),]
  sets <- mres$enrichment_result$set
  ridgedata <- lapply( sets , function(myset) {
    genes <- genesets[[which(names(genesets) == myset)]]
    generanks <- mres$ranked_profile[which(rownames(mres$ranked_profile) %in% genes),]
    return(generanks)
  })
  names(ridgedata) <- sets
  myrange <- range(mres$ranked_profile[,1])
  par(mar=c(5,20,3,2))

  vioplot(ridgedata,horizontal=TRUE,las=1,
    ylim=myrange,main="Gene sets of interest",
    xlab="differential gene rank")

  vioplot(ridgedata,horizontal=TRUE,las=1,
    ylim=myrange,col="white",main="Gene sets of interest",
    rectCol="gray", xlab="differential gene rank")
  beeswarm(ridgedata,add=TRUE,horizontal = TRUE, pch=19,cex=1,corral = "gutter")
})


```

Now make some heatmaps

```{r,heat_customgs}

sets <- names(custom_sets)

z <- lapply(sets, function(set){
  genes <- genesets[[which(names(genesets) == set)]]
  refgenes <- sapply(strsplit(rownames(rpm1)," "),"[[",2)
  hm <- as.matrix(rpm1[which(refgenes %in% genes),])
  hm <- hm/rowMeans(hm)
  colfunc <- colorRampPalette(c("blue", "white", "red"))
  heatmap.2(hm,trace="none",col=colfunc(25),scale="row", margins = c(6,15), cexRow=0.9,
    cexCol=0.9, main=paste(set) )
})

z <- lapply(sets, function(set){
  genes <- genesets[[which(names(genesets) == set)]]
  refgenes <- sapply(strsplit(rownames(rpm)," "),"[[",2)
  hm <- as.matrix(rpm[which(refgenes %in% genes),])
  hm <- hm/rowMeans(hm)
  hm[is.na(hm)] <- 0
  colfunc <- colorRampPalette(c("blue", "white", "red"))
  heatmap.2(hm,trace="none",col=colfunc(25),scale="row", margins = c(6,15), cexRow=0.7,
    cexCol=0.9, main=paste(set) )
})


refgenes <- sapply(strsplit(rownames(rpm1)," "),"[[",2)
gshm <- lapply(sets, function(set){
  genes <- genesets[[which(names(genesets) == set)]]
  hm <- as.matrix(rpm1[which(refgenes %in% genes),])
  hm <- hm/rowMeans(hm)
  return(colMeans(hm))
})

gshm <- do.call(rbind,gshm)
rownames(gshm) <- sets

colfunc <- colorRampPalette(c("blue", "white", "red"))
heatmap.2(gshm,trace="none",col=colfunc(25),scale="row", margins = c(6,15), cexRow=0.7,
  cexCol=0.9, main="Gene set heatmap" )

refgenes <- sapply(strsplit(rownames(rpm)," "),"[[",2)
gshm <- lapply(sets, function(set){
  genes <- genesets[[which(names(genesets) == set)]]
  hm <- as.matrix(rpm[which(refgenes %in% genes),])
  hm <- hm/rowMeans(hm)
  hm[is.na(hm)] <- 0
  return(colMeans(hm))
})

gshm <- do.call(rbind,gshm)
rownames(gshm) <- sets

colfunc <- colorRampPalette(c("blue", "white", "red"))
heatmap.2(gshm,trace="none",col=colfunc(25),scale="row", margins = c(6,15), cexRow=0.7,
  cexCol=0.9, main="Gene set heatmap" )

```

Mitch multi contrasts with custom gene sets

```{r,mitch_custom2}

mm <- list("KO"=dge1,"drug"=dge4)
m <- mitch_import(mm, DEtype="deseq2",geneTable=gt)

mres <- mitch_calc(m, custom_sets, priority="effect")
head(mres$enrichment_result,20) %>%
  kbl(caption = "Top mito pathway differences in contrast 1 and 4") %>% kable_paper("hover", full_width = F)
#mitch_report(mres,outfile="mitchmulti_effect_mito.html",overwrite=TRUE)
write.table(mres$enrichment_result,file="mitch_multi_effect_mito.tsv",quote=FALSE,sep="\t",row.names=FALSE)
mx <- as.matrix(mres$enrichment_result[,4:5])
rownames(mx) <- mres$set
rownames(mx) <- mres$enrichment_result$set

colfunc <- colorRampPalette(c("blue", "white", "red"))

heatmap.2(mx,trace="none",col=colfunc(25),scale="none", margins = c(5,22), cexRow=1,
    cexCol=0.9, main=paste("Top ranked Reactomes") )


mres <- mitch_calc(m, custom_sets, priority="SD")
head(mres$enrichment_result,20) %>% kbl(caption = "Top mito pathway differences in contrast 1 and 4") %>% kable_paper("hover", full_width = F)
#mitch_report(mres,outfile="mitchmulti_discord_mito.html",overwrite=TRUE)
write.table(mres$enrichment_result,file="mitch_multi_discord_mito.tsv",quote=FALSE,sep="\t",row.names=FALSE)

mx <- as.matrix(mres$enrichment_result[,4:5])
rownames(mx) <- mres$set
rownames(mx) <- mres$enrichment_result$set

colfunc <- colorRampPalette(c("blue", "white", "red"))

heatmap.2(mx,trace="none",col=colfunc(25),scale="none", margins = c(5,22), cexRow=1,
    cexCol=0.9, main=paste("Top ranked Reactomes") )

```

## Conclusion

TODO

## Session information

For reproducibility.

```{r,sessioninfo}

sessionInfo()

```
